### ISSUE TITLE
Miscellaneous issues with KKBSP v5

### ISSUE THREAD

**Issue by ilcane87** [FlailingFog/KK-Blender-Shader-Pack#65]
This picture was taken after the "Convert and Apply colors " step, and shows two small problems:

![image](https://user-images.githubusercontent.com/34897042/141003819-deec79e6-c492-43c5-b37a-98df59c0bb71.png)

1. The right eyewhite is assigned a hair material, which is easily fixed by merging it with the Eyewhites (sirome) material.

2. One vertex of the lower lip was merged with one of the upper lip; I verified that this doesn't happen when importing with CATS using  an increased scale, so I'm assuming the "Finalize PMX" step performs a Merge By Distance on all vertices, which is a risky operation because almost-overlapping vertices may get joined together too, especially on small scaled high poly models.

The problem with the increased scale workaround is that some bones get weird resizing results like these: 

![image](https://user-images.githubusercontent.com/34897042/141004832-190971ff-3445-427d-a871-859988e02a7c.png)

Either removing the merge operation or fixing the bone resizing would solve the issue.

Here's the reference character, I used the default outfit:

https://db.bepis.io/koikatsu/view/17005

On a side note, is there a way to revert the permanent Wireframe looks (X-Ray looks in edit mode) of bones in the imported armature from CATS?
I tried all of the typical functions and overlay options, but none of them seem to affect them:

![image](https://user-images.githubusercontent.com/34897042/141008291-2c380b13-3894-4472-9d66-3f4b42c494ac.png)
![image](https://user-images.githubusercontent.com/34897042/141008184-7b81f840-f816-4485-acd4-3d4227d30138.png)

**Comment by FlailingFog** [comment]
> 1. The right eyewhite is assigned a hair material, which is easily fixed by merging it with the Eyewhites (sirome) material.

I added a fix for this yesterday, does that one work or is it still doing it?


> 2. so I'm assuming the "Finalize PMX" step performs a Merge By Distance on all vertices,

It only does this for the body materials to prevent seams from appearing on the neck / down the torso. The merge distance can be set (in separatebody.py line 147) so I can check the card and set it lower to a value that works.

>On a side note, is there a way to revert the permanent Wireframe looks (X-Ray looks in edit mode) of bones in the imported armature from CATS?

It looks like CATS changes the viewport display options in this tab to Wire
![image](https://user-images.githubusercontent.com/65811931/141027379-065a32b3-c2d5-4edc-a95c-46745f868b91.png)

**Comment by MediaMoots** [comment]
@ilcane87 
>  importing with CATS using an increased scale

How did you import? 
With this button:
![image](https://user-images.githubusercontent.com/81146974/141041227-23480b7b-99da-4bec-835c-748fa4ff4d63.png)

Or this button?
![image](https://user-images.githubusercontent.com/81146974/141041271-c708cf67-e011-43c1-92c4-821197227a51.png)

If using the second one, try using the first one.

**Comment by ilcane87** [comment]
@MediaMoots 

I tried both. The Import Model button results in a small scaled character (less than 1.5m of height, which I assume is the original height), and presents the vertex merge issue after the "Finalize PMX" step.
Then I tried using the MMD button and choosing the default scale of 0.08 in the custom settings, which resulted in a character of roughly the same height of the previous try, which still presents the issue after the following step.
Then I tried using the MMD button with an increased scale of 0.4, which resulted in a character taller than 6m, and that one was safe after the following step.

Note that even in the small scaled cases the mouth shape-keys work correctly (without the vertex issue) before performing the "Finalize PMX" step, so the issue isn't with CATS or KKBP_Exporter:

![image](https://user-images.githubusercontent.com/34897042/141055897-9b8be04d-a942-4840-abe3-862b5e2f811f.png)

**Comment by ilcane87** [comment]
@FlailingFog 

> I added a fix for this yesterday, does that one work or is it still doing it?

Just cloned the repository again and it's fixed, well done.

> It only does this for the body materials to prevent seams from appearing on the neck / down the torso. The merge distance can be set (in separatebody.py line 147) so I can check the card and set it lower to a value that works.

I tried adding one 0 and the issue didn't appear:

_bpy.ops.mesh.remove_doubles(threshold=0.00001)_

> It looks like CATS sets the viewport display options in this tab to Wire

Completely forgot about that tab, thanks a lot!


Random question: I noticed that aside from the Template Face material, all the materials belonging to face components are set to use the Template Body Outline material rather than the Face one for the solidify modifier, is that intended?

![image](https://user-images.githubusercontent.com/34897042/141062079-084b1fd0-9597-410f-9620-1b354f36bf3a.png)

There's slightly different results when switching between the two, not that I would've normally noticed if I didn't check the setup.

**Comment by MediaMoots** [comment]
> @ilcane87
> 
> > importing with CATS using an increased scale
> 
> How did you import? With this button: ![image](https://user-images.githubusercontent.com/81146974/141041227-23480b7b-99da-4bec-835c-748fa4ff4d63.png)
> 
> Or this button? ![image](https://user-images.githubusercontent.com/81146974/141041271-c708cf67-e011-43c1-92c4-821197227a51.png)

Nice, thanks!

**Comment by MediaMoots** [comment]
> @MediaMoots
> 
> I tried both. The Import Model button results in a small scaled character (less than 1.5m of height, which I assume is the original height), and presents the merged vertices issue after the "Finalize PMX" step. Then I tried using the MMD button and choosing the default scale of 0.08 in the custom settings, which resulted in a character of roughly the same height of the previous try, which still presents the issue after the following step. Then I tried using the MMD button with an increased scale of 0.4, which resulted in a character taller than 6m, and that one was safe after the following step.
> 
> Note that even in the small scaled cases the mouth shape-keys work correctly (without the merged vertices) before performing the "Finalize PMX" step, so the issue isn't with CATS or KKBP_Exporter:
> 
> ![image](https://user-images.githubusercontent.com/34897042/141055897-9b8be04d-a942-4840-abe3-862b5e2f811f.png)

Nice, thanks!

**Comment by FlailingFog** [comment]
> Random question:

The index should start at zero, so that's why it's set to 11.
I also added the extra zero to the threshold value, thanks for checking it

**Comment by ilcane87** [comment]
> The index should start at zero, so that's why it's set to 11. I also added the extra zero to the threshold value, thanks for checking it

The index is calculated starting from the current index of each material, so:

- Template Face is at slot 0, so the modifier will use slot 0 + 11 = 11, which is Template Face Outline
- Template Eyebrows is at slot 1, so the modifier will use slot 1 + 11 = 12, which is Template **Body** Outline
- Template Nose is at slot 2, so the modifier will use slot 2 + 11 = 13, which doesn't exist, so it falls back to the last slot in the list, which is again Template **Body** Outline
- Etc...

In other words, all of the face elements are mapped to use the wrong (I assume) outline.

**Comment by FlailingFog** [comment]
Sorry I misread your point. I guess you can call it intended... I've known about this but I personally use freestyle instead of the outline modifier so I was kind of just going to leave it like that until someone said something. I'll reorder the materials in the scripts

The outline index for the clothes objects used to work, I just haven't looked into it yet.

**Comment by ilcane87** [comment]
> freestyle

I'd been considering using that, does it give you as much control as the modifier for outlines?
For instance, can you make the lines more or less pronounced based on vertex groups?

**Comment by FlailingFog** [comment]
I only use it as an outline on/off type thing so I don't know if you can finetune it like that

**Comment by FlailingFog** [comment]
I don't know what happened but the automatic clothes material offset is now working for me in both 2.93.4 and 3.0.0 so I'm going to close this. Let me know if it's still happening.

**Comment by MediaMoots** [comment]
@FlailingFog Does the automatic color import system still work on Blender 3.0?

**Comment by FlailingFog** [comment]
> Does the automatic color import system still work on Blender 3.0?

Yep, I didn't see a difference between 2.93 and 3.0
